### ISSUE TITLE
Partial Evaluation Return Residuals with Missing Entities

### ISSUE THREAD

**Issue by andrewmwells-amazon** [cedar-policy/cedar#312]
### Category

Cedar language features, User level API changes

### Describe the feature you'd like to request

From this slack thread:
https://cedar-policy.slack.com/archives/C0547KH7R19/p1695072247582719

It could be useful for partial evaluation to return residuals indicating missing entities. For example, consider the policy:
`permit(principal in Group::"B", action, resource);`
With schema:
```
{
    "": {
      "entityTypes": {
        "Group" : {
            "memberOfTypes" : [ "Group" ]
        },
        "User" : {
          "memberOfTypes" : [ "Group" ],
          "shape": {
            "type": "Record",
            "attributes": {
            }
        }
      }
    },
    "actions": {
      "view" : {
        "appliesTo" : {
          "resourceTypes" : [ "User" ],
          "principalTypes" : [ "User" ]
        },
        "memberOf" : [ ]
      }
    }
  }
}
```
and the input entities:
```
[
    {
        "uid": {
            "type": "User",
            "id": "bob"
        },
        "attrs": {},
        "parents":[{"type": "Group", "id": "A"}]
    }
]
```

Where it is assumed the complete entities file is:
```
[
    {
        "uid": {
            "type": "User",
            "id": "bob"
        },
        "attrs": {},
        "parents":[{"type": "Group", "id": "A"}]
    },
    {
        "uid": {
            "type": "Group",
            "id": "A"
        },
        "attrs": {},
        "parents":[ {"type": "Group", "id": "B"} ]
    },
    {
        "uid": {
            "type": "Group",
            "id": "B"
        },
        "attrs": {},
        "parents":[ ]
    }

]
```

On partially evaluating, we'd like to get some answer like: 
`would be allowed if A in B, but I don't know what A is`

### Describe alternatives you've considered

Maintaining the current behavior.

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [X] ⚠️ This feature might incur a breaking change

**Comment by cdisselkoen** [comment]
Partial evaluation does provide a `Entities::partial()` method (see [here](https://github.com/cedar-policy/cedar/blob/511391e57213a2dc372dbbd2943afeb3109e2ea8/cedar-policy/src/api.rs#L226)).  For a partial `Entities` created with that method, missing entities return residuals.  Does that solve your problem, or am I misunderstanding?

**Comment by prakol16** [comment]
Can we close in favor of #313 ?

**Comment by 0x00A5** [comment]
> Can we close in favor of #313 ?

I have a similar request about partial evaluation returning missing entities. Please let me know if I should file a separate issue for it.

If I give the following entity file to the evaluator,

```
[
    {
        "uid": {
            "type": "User",
            "id": "alice"
        },
        "attrs": {
            "referrer": {
                "__entity": {
                    "type": "User",
                    "id": "bob"
                }
            }
        },
        "parents":[]
    }
]
```

and the complete entity file is

```
[
    {
        "uid": {
            "type": "User",
            "id": "alice"
        },
        "attrs": {
            "referrer": {
                "__entity": {
                    "type": "User",
                    "id": "bob"
                }
            }
        },
        "parents":[]
    },
    {
        "uid": {
            "type": "User",
            "id": "bob"
        },
        "attrs": {
            "num_of_invitation_code": 1
        },
        "parents":[]
    }
]
```

when partially evaluating the following policy:

```
permit(
  principal,
  action,
  resource
)
when {
  principal.referrer.num_of_invitation_code > 0
};
```

I'd like the evaluation response to tell me that `User::"bob"` is a missing entity via the `EvalautionErrorKind` in the diagnostics, but the evaluator currently doesn't give any diagnostics when using a partial store and give a concrete `Deny` when using a regular store.

**Comment by prakol16** [comment]
I guess what you are saying is that when using a partial store you cannot know which entities have become residuals? I think you should file a new issue, but in my opinion the right framing is to ask to [expose the `unknowns` method](https://github.com/cedar-policy/cedar/blob/511391e57213a2dc372dbbd2943afeb3109e2ea8/cedar-policy-core/src/ast/expr.rs#L476) on a residual policy so that you can get all the unknowns (which correspond to the residual entities). The errors are typically reserved for true errors that prevent evaluation, and a missing entity in a partial entity store would not be a true error.

**Comment by cdisselkoen** [comment]
Closed by #353 (feel free to reopen if you disagree)
